software_bundle/software_bundle.proto

```diff
@@ -0,0 +1,229 @@
+  //                              (with acknowledgement = INSTALL_COMPLETE)
+  // STREAM END
+  // Note that all installations will install the software bundle, but that
+  // software bundle will not be applied until the affected services
+  // are reloaded.
+  // To reload services after installing software bundles, issue a
+  // FinalizeSoftwareBundles RPC after finishing the InstallSoftwareBundle RPC.
+  rpc InstallSoftwareBundle(stream InstallSoftwareBundleRequest)
+      returns (stream InstallSoftwareBundleResponse);
+
+  // UninstallSoftwareBundle uninstalls an software bundle.
+  // It returns an error if the requested software bundle is not installed.
+  rpc UninstallSoftwareBundle(UninstallSoftwareBundleRequest)
+      returns (UninstallSoftwareBundleResponse);
+}
+
+message ListSoftwareBundlesRequest{
+}
+message ListSoftwareBundlesResponse{
+  enum InstallStatus {
+    // UNSPECIFIED should be treated as an error.
+    UNSPECIFIED = 0;
+    INSTALLED = 1;
+    NOT_INSTALLED = 2;
+
+    // This software bundle was force-installed - checks regarding
+    // signature or dependencies may have been ignored
+    FORCE_INSTALLED = 3;
+  }
+  message SoftwareBundle {
+    // name of the software bundle
+    string software_bundle_name = 1;
+    // arbitrarily-formatted version string for the software bundle.
+    // This version would be extracted from the software bundle itself,
+    // as a way for vendors to identify extensions.
+    // (since the software_bundle_name is provided by the client).
+    string version = 2;
+    string description = 3;
+    InstallStatus install_status = 4;
+
+    // If persistent is true, then the system will ensure
+    // that this software bundle is applied on startup
+    // if system reboots (e.g. via the System.Reboot RPC).
+    // If persistent is false then the system will not
+    // re-apply this software bundle after startup.
+    bool persistent = 5;
+
+    // bool indicating if this software bundle has a valid signature
+    bool signature_valid = 6;
+
+    // available is true if the software bundle file is still available
+    // on the system.
+    // the software bundle may be installed but no longer available
+    // if the software bundle file has been removed.
+    bool available = 7;
+
+    // any errors attached to this software bundle
+    string errors = 8;
+
+    // affected_processes is a list of the processes which can be affected
+    // by this software bundle.
+    // This list may be incomplete if the system cannot determine all of the
+    // processes that could be affected.
+    repeated string affected_processes = 9;
+
+    // processes_pending_restart is a list of the processes which are waiting
+    // be restarted to apply the software bundle.
+    // This list will be cleared after a FinalizeSoftwareBundles RPC executes
+    // and restarts the affected processes.
+    repeated string processes_pending_restart = 10;
```

Comment:
How do we know the operational state of the processes, related to the software bundles?      

If the process isn't pending restart, does that mean the installed bundle is active?   

maybe a direct indication would be helpful.  Like
```
  enum ActivationStatus {
    // UNSPECIFIED should be treated as an error.
    UNSPECIFIED = 0;
    NOT_ACTIVATED = 1;
    ACTIVATED = 2;
  }
```